### PR TITLE
Add the palette-picker inside the timetable balloons

### PR DESCRIPTION
**Pull request by florv**
This allow to change the color of breaks and session blocks from the balloons directly.

### CODE REVIEW

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html [hunk 1]
```diff
@@ -33,3 +33,18 @@
...
+{% macro render_palette_picker(obj, endpoint, color_list) %}
+    <dl class="i-data-list thirty-seventy">
+        <dt>{% trans %}Color{% endtrans %}</dt>
+        <dd>
+            <a class="i-button icon-palette palette-picker palette-picker-trigger"
+               style="{{ obj.colors.css }}"
+               data-method="PATCH"
+               data-href="{{ url_for(endpoint, obj) }}"
+               data-palette={{ color_list | tojson }}
+               data-initial-color={{ obj.colors.css | tojson}}
```
Comment (mic4ael) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: missing space after `tojson`

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html [hunk 2]
```diff
@@ -33,3 +33,18 @@
...
+
+{% macro render_palette_picker(obj, endpoint, color_list) %}
+    <dl class="i-data-list thirty-seventy">
+        <dt>{% trans %}Color{% endtrans %}</dt>
+        <dd>
+            <a class="i-button icon-palette palette-picker palette-picker-trigger"
+               style="{{ obj.colors.css }}"
+               data-method="PATCH"
+               data-href="{{ url_for(endpoint, obj) }}"
+               data-palette={{ color_list | tojson }}
```
Comment (mic4ael) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: missing `"`

File: indico/htdocs/js/indico/jquery/palettepicker.js [hunk 3]
```diff
@@ -111,7 +112,12 @@
                 events: {
                     show: self._updateSelection.bind(self)
                 }
-            });
+            }
+            if (self.options.qtipConstructor) {
+                self.options.qtipConstructor(element, qtipOptions);
```
Comment (mic4ael) [inline indico/htdocs/js/indico/jquery/palettepicker.js]: what do you think about using `.call(element, qtipOptions)` and accessing the element as `this` inside the `qtipConstructor`?

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html (hunk 1 above)
Comment (ThiefMaster) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: also missing quotes for `data-palette`

File: indico/htdocs/js/indico/jquery/qbubble.js [hunk 4]
```diff
@@ -41,8 +41,29 @@
             var self = this;
             var classes = self.options.style ? self.options.style.classes : '';
 
+            var originalHideCallback = self.options.events && self.options.events.hide;
+            var originalFocusCallback = self.options.events && self.options.events.focus;
+            var hideCallback = function(evt, api) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/qbubble.js]: you could just use `function hideCallback` - no need for the `var hideCallback = ...` notation

File: indico/modules/events/timetable/controllers/manage.py [hunk 5]
```diff
@@ -145,3 +147,21 @@ def _checkParams(self, params):
     def _process(self):
         html = render_entry_info_balloon(self.entry, editable=True, sess=self.session)
         return jsonify(html=html)
+
+
+class RHBreakREST(RHManageTimetableBase):
+
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/controllers/manage.py]: no empty line unless there's a docstring

File: indico/modules/events/timetable/models/breaks.py [hunk 6]
```diff
@@ -81,6 +82,10 @@ def end_dt(self):
     def __repr__(self):
         return format_repr(self, 'id', _text=self.title)
 
+    @locator_property
+    def locator(self):
+        return dict(self.timetable_entry.locator, break_id=self.id)
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/models/breaks.py]: `self.timetable_entry.event_new.locator` - you don't want the timetable entry id in this locator

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html (hunk 2 above)
Comment (florv) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: With the quotes I'll get this something like this:
```
data-initial-color=""color: #202020 !important; background: #dfe555 !important""
```
which won't be parsed as wanted by the browser.

Comment (mic4ael) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: ahh, do you need to call `tojson` filter in that case?

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html (hunk 1 above)
Comment (florv) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: If I put quotes around `{{ color_list | tojson }}`, how do I then deal with the fact that `| tojson` is going to print quotes?

Comment (florv) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: I found out that by using `| tojson | forceescape` I get a properly escaped output.

File: indico/htdocs/js/indico/jquery/palettepicker.js (hunk 3 above)
Comment (florv) [inline indico/htdocs/js/indico/jquery/palettepicker.js]: My personal preference is to have an explicit argument but if someone else think using `this` is clearer I'll change it.

File: indico/htdocs/js/indico/jquery/qbubble.js [hunk 7]
```diff
@@ -41,8 +41,29 @@
...
+            var originalHideCallback = self.options.events && self.options.events.hide;
+            var originalFocusCallback = self.options.events && self.options.events.focus;
+            function hideCallback(evt, api) {
+                if (self._hasNestedOpen) {
+                    evt.preventDefault();
+                } else if (originalHideCallback) {
+                    originalHideCallback(evt, api);
+                }
+            };
+            var focusCallback = function(evt, api) {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/qbubble.js]: `function focusCallback`

File: indico/htdocs/js/indico/jquery/qbubble.js [hunk 8]
```diff
@@ -41,8 +41,29 @@
...
 
+            var originalHideCallback = self.options.events && self.options.events.hide;
+            var originalFocusCallback = self.options.events && self.options.events.focus;
+            function hideCallback(evt, api) {
+                if (self._hasNestedOpen) {
+                    evt.preventDefault();
+                } else if (originalHideCallback) {
+                    originalHideCallback(evt, api);
+                }
+            };
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/jquery/qbubble.js]: no semicolon

File: indico/htdocs/js/indico/modules/timetable/timetable/Draw.js [hunk 9]
```diff
@@ -649,6 +649,31 @@ function loadBalloonContent(self, api, editable) {
...
+                    $.ajax({
+                        url: $picker.data('href'),
+                        method: $picker.data('method'),
+                        data: JSON.stringify({'colors': {'text': text, 'background': background}}),
+                        dataType: 'json',
+                        contentType: 'application/json',
+                        error: handleAjaxError,
+                        complete: function() {
+                            self.setColors(text, background);
+                            IndicoUI.Dialogs.Util.progress()();
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/timetable/timetable/Draw.js]: That doesn't do what you want - the point of passing `complete: IndicoUI.Dialogs.Util.progress()` is that the first invocation happens before the request is sent and then the returned function to close the dialog is executed when the request finished. So in this case you need to do call it before the `$.ajax()` call and assign the returned function and then call that one in here.

File: indico/modules/events/timetable/blueprint.py [hunk 10]
```diff
@@ -44,6 +45,7 @@
 _bp.add_url_rule('/manage/timetable/', 'timetable_rest', RHTimetableREST, methods=('POST',))
 _bp.add_url_rule('/manage/timetable/<int:entry_id>', 'timetable_rest', RHTimetableREST, methods=('PATCH', 'DELETE'))
 _bp.add_url_rule('/manage/timetable/session/<int:session_id>/', 'manage_session', RHManageSessionTimetable)
+_bp.add_url_rule('/manage/timetable/break/<int:break_id>/', 'break_rest', RHBreakREST, methods=('PATCH',))
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/blueprint.py]: no trailing slash here

File: indico/modules/events/timetable/models/breaks.py [hunk 11]
```diff
@@ -81,6 +82,10 @@ def end_dt(self):
     def __repr__(self):
         return format_repr(self, 'id', _text=self.title)
 
+    @locator_property
+    def locator(self):
+        return dict(confId=self.timetable_entry.event_new.id, break_id=self.id)
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/models/breaks.py]: `dict(self.timetable_entry.event_new, break_id=self.id)` or `dict(confId=self.timetable_entry.event_id, break_id=self.id)`

File: indico/modules/events/timetable/templates/balloons/_balloon_utils.html [hunk 12]
```diff
@@ -33,3 +33,19 @@
...
+
+{% macro render_palette_picker(obj, endpoint, color_list) %}
+    <dl class="i-data-list thirty-seventy">
+        <dt>{% trans %}Color{% endtrans %}</dt>
+        <dd>
+            <a class="i-button icon-palette palette-picker palette-picker-trigger"
+               style="{{ obj.colors.css }}"
+               data-method="PATCH"
+               data-href="{{ url_for(endpoint, obj) }}"
+               {# forceescape needed to properly handle the double quotes inside |tojson's output #}
```
Comment (ThiefMaster) [inline indico/modules/events/timetable/templates/balloons/_balloon_utils.html]: a bit verbose? you always need to `|forceescape` when using `tojson` outside `<script>` tags and in all the other places we don't have such a comment either

File: indico/htdocs/sass/partials/_icons.scss [hunk 13]
```diff
@@ -61,6 +61,32 @@ i[class^="icon-"]:before {
     opacity: 0.15;
 }
 
+/* How to add icons
+ *
+ *  1. Login into icomoon app https://icomoon.io/app/
+ *  2. Remove any existing project (Top-left hamburger icon, "Manage projects", "Remove" button)
+ *  3. Import "sejection.js" from Indico repository (Top-left hamburger icon, "Import project",
```
Comment (ThiefMaster) [inline indico/htdocs/sass/partials/_icons.scss]: `selection.json`

File: indico/htdocs/js/indico/modules/timetable/timetable/Draw.js [hunk 14]
```diff
@@ -644,6 +658,35 @@ function loadBalloonContent(self, api, editable) {
...
+                selectedColor: $picker.data('initial-color'),
+                onSelect: function(background, text) {
+                    var killProgress = IndicoUI.Dialogs.Util.progress();
+                    $.ajax({
+                        url: $picker.data('href'),
+                        method: $picker.data('method'),
+                        data: JSON.stringify({'colors': {'text': text, 'background': background}}),
+                        dataType: 'json',
+                        contentType: 'application/json',
+                        error: function() {
```
Comment (ThiefMaster) [inline indico/htdocs/js/indico/modules/timetable/timetable/Draw.js]: You can simplify this:
```
complete: IndicoUI.Dialogs.Util.progress(),
error: handleAjaxError,
```

File: indico/htdocs/sass/partials/_icons.scss [hunk 15]
```diff
@@ -61,6 +61,35 @@ i[class^="icon-"]:before {
     opacity: 0.15;
 }
 
+/* How to add icons
+ *
+ *  1. Login into icomoon app https://icomoon.io/app/
+ *  2. Remove any existing project (Top-left hamburger icon, "Manage projects", "Remove" button)
+ *  3. Import "sejection.json" from Indico repository (Top-left hamburger icon, "Import project",
```
Comment (ThiefMaster) [inline indico/htdocs/sass/partials/_icons.scss]: seLection

**Pull request merged**